From 9f42776b045452ebffc9ca5f2d8b771f62f3acf4 Mon Sep 17 00:00:00 2001
From: Haibo Chen <haibo.chen@nxp.com>
Date: Thu, 14 Jun 2018 17:06:06 +0800
Subject: [PATCH 4015/5242] MLK-18612-3 iio: adc: imx8qxp_adc: fix register
 setting

commit  c9feec063864d0840d87a7bbb5a23f324f5e63c1 from
https://source.codeaurora.org/external/imx/linux-imx.git

For the register CFG/TCTRLx/CMDLx/CMDHx, all the defined bits of
these register are setting, so no need to read the register
first, and use "|=", and will cause issue, the bit setting will be
impact by the last time setting. So write these register directlly.

Signed-off-by: Haibo Chen <haibo.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/iio/adc/imx8qxp_adc.c |   12 ++++--------
 1 file changed, 4 insertions(+), 8 deletions(-)

diff --git a/drivers/iio/adc/imx8qxp_adc.c b/drivers/iio/adc/imx8qxp_adc.c
index 331e241..74da438 100644
--- a/drivers/iio/adc/imx8qxp_adc.c
+++ b/drivers/iio/adc/imx8qxp_adc.c
@@ -267,8 +267,7 @@ static void imx8qxp_adc_reg_config(struct imx8qxp_adc *adc)
 	u32 adc_cfg, adc_tctrl, adc_cmdl, adc_cmdh;
 	u32 t_id, c_id;
 
-	adc_cfg = readl(adc->regs + IMX8QXP_REG_ADC_CFG);
-	adc_cfg |= adc->adc_cfg.pwren << IMX8QXP_REG_ADC_CFG_PWREN_SHIFT
+	adc_cfg = adc->adc_cfg.pwren << IMX8QXP_REG_ADC_CFG_PWREN_SHIFT
 			| adc->adc_cfg.pudly << IMX8QXP_REG_ADC_CFG_PUDLY_SHIFT
 			| adc->adc_cfg.refsel << IMX8QXP_REG_ADC_CFG_REFSEL_SHIFT
 			| adc->adc_cfg.pwrsel << IMX8QXP_REG_ADC_CFG_PWRSEL_SHIFT
@@ -276,24 +275,21 @@ static void imx8qxp_adc_reg_config(struct imx8qxp_adc *adc)
 	writel(adc_cfg, adc->regs + IMX8QXP_REG_ADC_CFG);
 
 	t_id = adc->trigger_id;
-	adc_tctrl = readl(adc->regs + IMX8QXP_REG_ADC_TCTRL0 + t_id * 4);
-	adc_tctrl |= adc->adc_trigger_ctrl[t_id].tcmd << IMX8QXP_REG_ADC_TCTRL_TCMD_SHIFT
+	adc_tctrl = adc->adc_trigger_ctrl[t_id].tcmd << IMX8QXP_REG_ADC_TCTRL_TCMD_SHIFT
 			| adc->adc_trigger_ctrl[t_id].tdly << IMX8QXP_REG_ADC_TCTRL_TDLY_SHIFT
 			| adc->adc_trigger_ctrl[t_id].tpri << IMX8QXP_REG_ADC_TCTRL_TPRI_SHIFT
 			| adc->adc_trigger_ctrl[t_id].hten << IMX8QXP_REG_ADC_TCTRL_HTEN_SHIFT;
 	writel(adc_tctrl, adc->regs + IMX8QXP_REG_ADC_TCTRL0 + t_id * 4);
 
 	c_id = adc->cmd_id - 1;
-	adc_cmdl = readl(adc->regs + IMX8QXP_REG_ADC_CMDL1 + c_id * 8);
-	adc_cmdl |= adc->adc_cmd_l[c_id].scale << IMX8QXP_REG_ADC_CMDL_CSCALE_SHIFT
+	adc_cmdl = adc->adc_cmd_l[c_id].scale << IMX8QXP_REG_ADC_CMDL_CSCALE_SHIFT
 			| adc->adc_cmd_l[c_id].mode << IMX8QXP_REG_ADC_CMDL_MODE_SHIFT
 			| adc->adc_cmd_l[c_id].diff << IMX8QXP_REG_ADC_CMDL_DIFF_SHIFT
 			| adc->adc_cmd_l[c_id].absel << IMX8QXP_REG_ADC_CMDL_ABSEL_SHIFT
 			| adc->adc_cmd_l[c_id].adch << IMX8QXP_REG_ADC_CMDL_ADCH_SHIFT;
 	writel(adc_cmdl, adc->regs + IMX8QXP_REG_ADC_CMDL1 + c_id * 8);
 
-	adc_cmdh = readl(adc->regs + IMX8QXP_REG_ADC_CMDH1 + c_id * 8);
-	adc_cmdh |= adc->adc_cmd_h[c_id].next << IMX8QXP_REG_ADC_CMDH_NEXT_SHIFT
+	adc_cmdh = adc->adc_cmd_h[c_id].next << IMX8QXP_REG_ADC_CMDH_NEXT_SHIFT
 			| adc->adc_cmd_h[c_id].loop << IMX8QXP_REG_ADC_CMDH_LOOP_SHIFT
 			| adc->adc_cmd_h[c_id].avgs << IMX8QXP_REG_ADC_CMDH_AVGS_SHIFT
 			| adc->adc_cmd_h[c_id].sts << IMX8QXP_REG_ADC_CMDH_STS_SHIFT
-- 
1.7.9.5

